File: internal/scheduler/validate_test.go
package scheduler_test

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/pseudomuto/skedge/internal/config"
	"github.com/pseudomuto/skedge/internal/scheduler"
)

// --- test helpers ---

func blk(subject, teacher, room string) scheduler.Block {
	return scheduler.Block{Name: subject, Teacher: teacher, Room: room}
}

func testDay(name string, blocks ...scheduler.Block) scheduler.DailySchedule {
	return scheduler.DailySchedule{Day: name, Blocks: blocks}
}

func testCohort(name string, days ...scheduler.DailySchedule) scheduler.Cohort {
	return scheduler.Cohort{Name: name, Schedule: days}
}

func testClass(name string, cohorts ...scheduler.Cohort) *scheduler.Class {
	return &scheduler.Class{Name: name, Cohorts: cohorts}
}

// weekFor builds a 5-day schedule where every day gets an independent copy of blocks.
func weekFor(blocks ...scheduler.Block) []scheduler.DailySchedule {
	dayNames := []string{"Monday", "Tuesday", "Wednesday", "Thursday", "Friday"}
	days := make([]scheduler.DailySchedule, len(dayNames))
	for i, name := range dayNames {
		cp := make([]scheduler.Block, len(blocks))
		copy(cp, blocks)
		days[i] = scheduler.DailySchedule{Day: name, Blocks: cp}
	}
	return days
}

// validSchedule returns a schedule that satisfies all validators when used with baseValidatorConfig.
// Grade8 / 8A: 5 days × 3 blocks (1 Math, 1 English, 1 Science each day).
// Smith (room 101) teaches all three subjects.
func validSchedule() []*scheduler.Class {
	days := weekFor(
		blk("Math", "Smith", "101"),
		blk("English", "Smith", "101"),
		blk("Science", "Smith", "101"),
	)
	return []*scheduler.Class{testClass("Grade8", testCohort("8A", days...))}
}

// baseValidatorConfig matches validSchedule.
func baseValidatorConfig() *config.Config {
	return &config.Config{
		Blocks: []string{"09:00", "10:00", "11:00"},
		Subjects: []config.Subject{
			{Name: "Math"}, {Name: "English"}, {Name: "Science"},
		},
		Classes: []config.Class{{
			Name:    "Grade8",
			Cohorts: []string{"8A"},
			Subjects: []config.ClassSubject{
				{Name: "Math", Blocks: 5},
				{Name: "English", Blocks: 5},
				{Name: "Science", Blocks: 5},
			},
		}},
		Teachers: []config.Teacher{{
			Name: "Smith",
			Room: "101",
			Subjects: []config.TeacherSubject{{
				Class:    "Grade8",
				Subjects: []string{"Math", "English", "Science"},
			}},
		}},
	}
}

// --- tests ---

func TestValidateNoTeacherConflict(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{
			name:     "valid",
			schedule: validSchedule(),
			wantErr:  false,
		},
		{
			name: "invalid: same teacher two cohorts same slot",
			schedule: []*scheduler.Class{testClass(
				"Grade8",
				testCohort("8A", testDay("Monday", blk("Math", "Smith", "101"))),
				testCohort("8B", testDay("Monday", blk("Math", "Smith", "101"))),
			)},
			wantErr: true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateNoTeacherConflict(tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

func TestValidateNoRoomConflict(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{
			name:     "valid",
			schedule: validSchedule(),
			wantErr:  false,
		},
		{
			name: "invalid: same room two cohorts same slot",
			schedule: []*scheduler.Class{testClass(
				"Grade8",
				testCohort("8A", testDay("Monday", blk("Math", "Smith", "101"))),
				testCohort("8B", testDay("Monday", blk("English", "Jones", "101"))), // different teacher, same room
			)},
			wantErr: true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateNoRoomConflict(tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

func TestValidateTeacherBlockLimit(t *testing.T) {
	t.Parallel()

	// buildWithSmithBlocks distributes n blocks to "Smith" across 2 cohorts (8A and 8B),
	// filling remaining slots with "Jones". Cohorts are kept to 3 blocks/day × 5 days = 15 slots each.
	buildWithSmithBlocks := func(n int) []*scheduler.Class {
		makeCohort := func(name string, smithCount int) scheduler.Cohort {
			var days []scheduler.DailySchedule
			placed := 0
			for _, d := range []string{"Monday", "Tuesday", "Wednesday", "Thursday", "Friday"} {
				var blocks []scheduler.Block
				for range 3 {
					if placed < smithCount {
						blocks = append(blocks, blk("Math", "Smith", "101"))
						placed++
					} else {
						blocks = append(blocks, blk("Math", "Jones", "102"))
					}
				}
				days = append(days, testDay(d, blocks...))
			}
			return testCohort(name, days...)
		}
		aCount := n / 2
		bCount := n - aCount
		return []*scheduler.Class{testClass("Grade8", makeCohort("8A", aCount), makeCohort("8B", bCount))}
	}

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{name: "valid: 21 blocks", schedule: buildWithSmithBlocks(21), wantErr: false},
		{name: "invalid: 22 blocks", schedule: buildWithSmithBlocks(22), wantErr: true},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateTeacherBlockLimit(tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

func TestValidateNoDuplicateSubjectPerDay(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{
			name:     "valid",
			schedule: validSchedule(),
			wantErr:  false,
		},
		{
			name: "invalid: Math appears twice on Monday",
			schedule: []*scheduler.Class{testClass(
				"Grade8",
				testCohort("8A", testDay(
					"Monday",
					blk("Math", "Smith", "101"),
					blk("Math", "Jones", "102"),
				)),
			)},
			wantErr: true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateNoDuplicateSubjectPerDay(tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

func TestValidateTeacherSubjects(t *testing.T) {
	t.Parallel()

	cfg := baseValidatorConfig()

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{
			name:     "valid",
			schedule: validSchedule(),
			wantErr:  false,
		},
		{
			name: "invalid: teacher teaching unauthorized subject",
			schedule: []*scheduler.Class{testClass(
				"Grade8",
				testCohort("8A", testDay(
					"Monday",
					blk("Art", "Smith", "101"), // Smith not authorized for Art
				)),
			)},
			wantErr: true,
		},
		{
			name: "invalid: teacher not authorized for class at all",
			schedule: []*scheduler.Class{testClass(
				"Grade8",
				testCohort("8A", testDay(
					"Monday",
					blk("Math", "Ghost", "999"), // Ghost not in config
				)),
			)},
			wantErr: true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateTeacherSubjects(cfg, tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

func TestValidateSubjectBlockCounts(t *testing.T) {
	t.Parallel()

	cfg := baseValidatorConfig()

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{
			name:     "valid: exact counts",
			schedule: validSchedule(),
			wantErr:  false,
		},
		{
			name: "invalid: too many Math blocks",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				// append an extra Math block on Friday (index 4)
				fri := &s[0].Cohorts[0].Schedule[4]
				fri.Blocks = append(fri.Blocks, blk("Math", "Smith", "101"))
				return s
			}(),
			wantErr: true,
		},
		{
			name: "invalid: too few English blocks",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				// remove the English block from Monday (index 0, block index 1)
				mon := &s[0].Cohorts[0].Schedule[0]
				mon.Blocks = append(mon.Blocks[:1], mon.Blocks[2:]...)
				return s
			}(),
			wantErr: true,
		},
		{
			name: "invalid: unknown subject in schedule",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				s[0].Cohorts[0].Schedule[0].Blocks[0].Name = "Alchemy"
				return s
			}(),
			wantErr: true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateSubjectBlockCounts(cfg, tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

func TestValidateNoGaps(t *testing.T) {
	t.Parallel()

	cfg := baseValidatorConfig() // expects 3 blocks per day

	tests := []struct {
		name     string
		schedule []*scheduler.Class
		wantErr  bool
	}{
		{
			name:     "valid",
			schedule: validSchedule(),
			wantErr:  false,
		},
		{
			name: "invalid: missing block on Monday",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				s[0].Cohorts[0].Schedule[0].Blocks = s[0].Cohorts[0].Schedule[0].Blocks[:2]
				return s
			}(),
			wantErr: true,
		},
		{
			name: "invalid: empty subject name",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				s[0].Cohorts[0].Schedule[0].Blocks[0].Name = ""
				return s
			}(),
			wantErr: true,
		},
		{
			name: "invalid: empty teacher name",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				s[0].Cohorts[0].Schedule[0].Blocks[0].Teacher = ""
				return s
			}(),
			wantErr: true,
		},
		{
			name: "invalid: empty room",
			schedule: func() []*scheduler.Class {
				s := validSchedule()
				s[0].Cohorts[0].Schedule[0].Blocks[0].Room = ""
				return s
			}(),
			wantErr: true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			err := scheduler.ValidateNoGaps(cfg, tc.schedule)
			if tc.wantErr {
				require.Error(t, err)
			} else {
				require.NoError(t, err)
			}
		})
	}
}

// TestValidate_combinedPassthrough ensures Validate calls all sub-validators.
func TestValidate_combinedPassthrough(t *testing.T) {
	t.Parallel()
	cfg := baseValidatorConfig()
	require.NoError(t, scheduler.Validate(cfg, validSchedule()))
}

// TestWithMaxRetries confirms the option is wired correctly: an infeasible config
// (no teachers) exhausts retries and returns an error.
func TestWithMaxRetries(t *testing.T) {
	t.Parallel()
	cfg := &config.Config{
		Blocks:   []string{"09:00"},
		Subjects: []config.Subject{{Name: "Math"}},
		Classes: []config.Class{{
			Name:     "Grade8",
			Cohorts:  []string{"8A"},
			Subjects: []config.ClassSubject{{Name: "Math", Blocks: 5}},
		}},
		Teachers: nil, // no teachers → no valid assignment possible
	}
	s := scheduler.New(cfg, scheduler.WithMaxRetries(3))
	_, err := s.Generate(t.Context())
	require.ErrorContains(t, err, fmt.Sprintf("%d", 3))
}

File: web/src/__tests__/scheduler/validate.test.ts
import { describe, expect, it } from 'vitest'

import type { Config } from '../../types/config'
import type { Block, DailySchedule, Cohort, ScheduleClass } from '../../types/schedule'

import {
  validate,
  validateNoGaps,
  validateNoDuplicateSubjectPerDay,
  validateNoRoomConflict,
  validateNoTeacherConflict,
  validateSubjectBlockCounts,
  validateTeacherBlockLimit,
  validateTeacherSubjects,
  MAX_TEACHER_BLOCKS_PER_WEEK,
} from '../../scheduler/validate'

function blk(name: string, teacher: string, room: string): Block {
  return { name, teacher, room }
}

function testDay(day: string, ...blocks: Block[]): DailySchedule {
  return { day, blocks }
}

function testCohort(name: string, ...schedule: DailySchedule[]): Cohort {
  return { name, schedule }
}

function testClass(name: string, ...cohorts: Cohort[]): ScheduleClass {
  return { name, cohorts }
}

function weekFor(...blocks: Block[]): DailySchedule[] {
  return ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'].map(day => ({
    day,
    blocks: blocks.map(b => ({ ...b })),
  }))
}

function validSchedule(): ScheduleClass[] {
  return [
    testClass(
      'Grade8',
      testCohort('8A', ...weekFor(
        blk('Math', 'Smith', '101'),
        blk('English', 'Smith', '101'),
        blk('Science', 'Smith', '101'),
      )),
    ),
  ]
}

function baseValidatorConfig(): Config {
  return {
    blocks: ['09:00', '10:00', '11:00'],
    subjects: [{ name: 'Math' }, { name: 'English' }, { name: 'Science' }],
    classes: [
      {
        name: 'Grade8',
        cohorts: ['8A'],
        subjects: [
          { name: 'Math', blocks: 5 },
          { name: 'English', blocks: 5 },
          { name: 'Science', blocks: 5 },
        ],
      },
    ],
    teachers: [
      {
        name: 'Smith',
        room: '101',
        subjects: [{ class: 'Grade8', subjects: ['Math', 'English', 'Science'] }],
      },
    ],
  }
}

describe('validateNoTeacherConflict', () => {
  it('passes for a valid schedule', () => {
    expect(validateNoTeacherConflict(validSchedule())).toBeNull()
  })

  it('catches same teacher in two cohorts at same slot', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort('8A', testDay('Monday', blk('Math', 'Smith', '101'))),
        testCohort('8B', testDay('Monday', blk('English', 'Smith', '102'))),
      ),
    ]
    const err = validateNoTeacherConflict(schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Smith')
    expect(err!.message).toContain('double-booked')
  })
})

describe('validateNoRoomConflict', () => {
  it('passes for a valid schedule', () => {
    expect(validateNoRoomConflict(validSchedule())).toBeNull()
  })

  it('catches same room used by two cohorts in the same slot', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort('8A', testDay('Monday', blk('Math', 'Smith', '101'))),
        testCohort('8B', testDay('Monday', blk('English', 'Jones', '101'))),
      ),
    ]
    const err = validateNoRoomConflict(schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('101')
    expect(err!.message).toContain('double-booked')
  })
})

describe('validateTeacherBlockLimit', () => {
  function buildWithSmithBlocks(count: number): ScheduleClass[] {
    const blocks: Block[] = []
    for (let i = 0; i < count; i++) {
      blocks.push(blk(`Subject${i}`, 'Smith', '101'))
    }
    // Distribute across 5 days, up to 5 blocks per day (25 total capacity)
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
    const schedule: DailySchedule[] = []
    let idx = 0
    for (const day of days) {
      const dayBlocks: Block[] = []
      for (let b = 0; b < 5 && idx < count; b++, idx++) {
        dayBlocks.push(blocks[idx])
      }
      if (dayBlocks.length > 0) schedule.push(testDay(day, ...dayBlocks))
    }
    return [testClass('Grade8', testCohort('8A', ...schedule))]
  }

  it(`passes for exactly ${MAX_TEACHER_BLOCKS_PER_WEEK} blocks`, () => {
    expect(validateTeacherBlockLimit(buildWithSmithBlocks(MAX_TEACHER_BLOCKS_PER_WEEK))).toBeNull()
  })

  it(`fails for ${MAX_TEACHER_BLOCKS_PER_WEEK + 1} blocks`, () => {
    const err = validateTeacherBlockLimit(buildWithSmithBlocks(MAX_TEACHER_BLOCKS_PER_WEEK + 1))
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Smith')
    expect(err!.message).toContain('exceeds weekly limit')
  })
})

describe('validateNoDuplicateSubjectPerDay', () => {
  it('passes for a valid schedule', () => {
    expect(validateNoDuplicateSubjectPerDay(validSchedule())).toBeNull()
  })

  it('catches Math appearing twice on Monday', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          testDay('Monday', blk('Math', 'Smith', '101'), blk('Math', 'Smith', '101')),
        ),
      ),
    ]
    const err = validateNoDuplicateSubjectPerDay(schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Math')
    expect(err!.message).toContain('Monday')
  })
})

describe('validateTeacherSubjects', () => {
  it('passes for a valid schedule', () => {
    expect(validateTeacherSubjects(baseValidatorConfig(), validSchedule())).toBeNull()
  })

  it('catches teacher teaching unauthorized subject', () => {
    const cfg = baseValidatorConfig()
    const schedule = [
      testClass(
        'Grade8',
        testCohort('8A', testDay('Monday', blk('Art', 'Smith', '101'))),
      ),
    ]
    const err = validateTeacherSubjects(cfg, schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Smith')
    expect(err!.message).toContain('Art')
  })

  it('catches teacher not in config for a class', () => {
    const cfg = baseValidatorConfig()
    const schedule = [
      testClass(
        'Grade8',
        testCohort('8A', testDay('Monday', blk('Math', 'Unknown', '101'))),
      ),
    ]
    const err = validateTeacherSubjects(cfg, schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Unknown')
    expect(err!.message).toContain('no subjects authorized')
  })
})

describe('validateSubjectBlockCounts', () => {
  it('passes when counts match exactly', () => {
    expect(validateSubjectBlockCounts(baseValidatorConfig(), validSchedule())).toBeNull()
  })

  it('fails when a subject has too many blocks', () => {
    const cfg = baseValidatorConfig()
    // 6 Math blocks instead of 5
    const schedule = [
      testClass(
        'Grade8',
        testCohort('8A', ...weekFor(
          blk('Math', 'Smith', '101'),
          blk('English', 'Smith', '101'),
          blk('Science', 'Smith', '101'),
        ), testDay('Extra', blk('Math', 'Smith', '101'))),
      ),
    ]
    const err = validateSubjectBlockCounts(cfg, schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Math')
  })

  it('fails when a subject has too few blocks', () => {
    const cfg = baseValidatorConfig()
    // Only 4 days of Math (missing one day)
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          ...days.map((day, i) =>
            testDay(
              day,
              i === 0
                ? blk('English', 'Smith', '101') // Monday: Math replaced with English
                : blk('Math', 'Smith', '101'),
              blk('English', 'Smith', '101'),
              blk('Science', 'Smith', '101'),
            ),
          ),
        ),
      ),
    ]
    const err = validateSubjectBlockCounts(cfg, schedule)
    expect(err).not.toBeNull()
    // Math will be 4 instead of 5, English will be 6 instead of 5
  })

  it('fails for an unknown subject', () => {
    // Config requires exactly 1 block each of Math/English/Science across all days.
    // The schedule satisfies those counts but adds an extra Art block - Art is not in config.
    const cfg: Config = {
      blocks: ['09:00', '10:00', '11:00', '11:30'],
      subjects: [{ name: 'Math' }, { name: 'English' }, { name: 'Science' }],
      classes: [
        {
          name: 'Grade8',
          cohorts: ['8A'],
          subjects: [
            { name: 'Math', blocks: 1 },
            { name: 'English', blocks: 1 },
            { name: 'Science', blocks: 1 },
          ],
        },
      ],
      teachers: [
        {
          name: 'Smith',
          room: '101',
          subjects: [{ class: 'Grade8', subjects: ['Math', 'English', 'Science'] }],
        },
      ],
    }
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          testDay(
            'Monday',
            blk('Math', 'Smith', '101'),
            blk('English', 'Smith', '101'),
            blk('Science', 'Smith', '101'),
            blk('Art', 'Smith', '101'),
          ),
        ),
      ),
    ]
    const err = validateSubjectBlockCounts(cfg, schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('Art')
  })
})

describe('validateNoGaps', () => {
  it('passes for a valid schedule', () => {
    expect(validateNoGaps(baseValidatorConfig(), validSchedule())).toBeNull()
  })

  it('fails when a day has a missing block', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          testDay('Monday', blk('Math', 'Smith', '101'), blk('English', 'Smith', '101')),
        ),
      ),
    ]
    const err = validateNoGaps(baseValidatorConfig(), schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('expected 3')
  })

  it('fails when a block has an empty subject', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          testDay(
            'Monday',
            blk('', 'Smith', '101'),
            blk('English', 'Smith', '101'),
            blk('Science', 'Smith', '101'),
          ),
        ),
      ),
    ]
    const err = validateNoGaps(baseValidatorConfig(), schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('empty field')
  })

  it('fails when a block has an empty teacher', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          testDay(
            'Monday',
            blk('Math', '', '101'),
            blk('English', 'Smith', '101'),
            blk('Science', 'Smith', '101'),
          ),
        ),
      ),
    ]
    const err = validateNoGaps(baseValidatorConfig(), schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('empty field')
  })

  it('fails when a block has an empty room', () => {
    const schedule = [
      testClass(
        'Grade8',
        testCohort(
          '8A',
          testDay(
            'Monday',
            blk('Math', 'Smith', ''),
            blk('English', 'Smith', '101'),
            blk('Science', 'Smith', '101'),
          ),
        ),
      ),
    ]
    const err = validateNoGaps(baseValidatorConfig(), schedule)
    expect(err).not.toBeNull()
    expect(err!.message).toContain('empty field')
  })
})

describe('validate', () => {
  it('passes for a fully valid schedule', () => {
    expect(validate(baseValidatorConfig(), validSchedule())).toBeNull()
  })
})
